stdlib/seq.mli

```diff
@@ -13,97 +13,807 @@
+   so as to understand the time and space requirements of sequences.
+
+   For the sake of simplicity, most of the documentation that follows
+   is written under the implicit assumption that the sequences at hand
+   are persistent.
+   We normally do not point out {i when} or {i how many times}
+   each function is invoked, because that would be too verbose.
+   For instance, in the description of [map], we write:
+   "if [xs] is the sequence [x0; x1; ...]
+    then [map f xs] is the sequence [f x0; f x1; ...]".
+   If we wished to be more explicit,
+   we could point out that the transformation takes place on demand:
+   that is, the elements of [map f xs] are computed only when they
+   are demanded. In other words,
+   the definition [let ys = map f xs] terminates immediately and
+   does not invoke [f]. The function call [f x0] takes place only when the
+   first element of [ys] is demanded, via the function call [ys()].
+   Furthermore, calling [ys()] twice causes [f x0] to be called twice
+   as well. If one wishes for [f] to be applied at most once to each
+   element of [xs], even in scenarios where [ys] is queried more than once,
+   then one should use [let ys = memoize (map f xs)].
+
+   As a general rule, the functions that build sequences, such as [map],
+   [filter], [scan], [take], etc., produce sequences whose elements are
+   computed only on demand. The functions that eagerly consume sequences,
+   such as [is_empty], [head], [tail], [find], [length], [iter], [fold_left],
+   etc., are the functions that force computation to take place.
+
+   When possible, we recommend using sequences rather than iterators
+   (functions of type [unit -> 'a option] that produce elements upon
+   demand). Whereas sequences can be persistent or ephemeral, iterators
+   are always ephemeral, and are typically more difficult to work with
+   than sequences. Two conversion functions, {!to_iterator} and
+   {!of_iterator}, are provided.
+
+    @since 4.07 *)
 
 type 'a t = unit -> 'a node
-(** The type of delayed lists containing elements of type ['a].
-    Note that the concrete list node ['a node] is delayed under a closure,
-    not a [lazy] block, which means it might be recomputed every time
-    we access it. *)
+(** A sequence [xs] of type ['a t] is a delayed list of elements of
+    type ['a]. Such a sequence is queried by performing a function
+    application [xs()]. This function application returns a node,
+    allowing the caller to determine whether the sequence is empty
+    or nonempty, and in the latter case, to obtain its head and tail. *)
 
 and +'a node =
   | Nil
   | Cons of 'a * 'a t (**)
-(** A fully-evaluated list node, either empty or containing an element
-    and a delayed tail. *)
+(** A node is either [Nil], which means that the sequence is empty,
+    or [Cons (x, xs)], which means that [x] is the first element
+    of the sequence and that [xs] is the remainder of the sequence. *)
+
+(** {1 Consuming sequences} *)
```

Comment:
Having an overview section at the beginning is good. But it'd be better if the information that some function consumes its input or not and to what extent were repeated in each function's docstring. It's verbose overall, but it's useful for when quickly checking the documentation of one specific function (e.g., via a "jump to" in merlin or some other IDE method).